File: 1463_Make_1/main.cpp
#include <iostream>
using namespace std;
int main() {
    int N;
    cin >> N;
    int i = N-1;
    int *a = new int[N+1];
    a[N] = 0;
    while(i > 0) {
        if(3*i<=N)
            a[i] = min(a[3*i],min(a[2*i],a[i+1])) + 1;
        else if(2*i<=N)
            a[i] = min(a[2*i],a[i+1]) + 1;
        else
            a[i] = a[i+1] + 1;
        i--;
    }
    cout << a[1];
    return 0;
}

File: 1463_Make_1/Main.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();

        int[] dp = new int[N+1];
        // dp[i]는 i를 만드는 데 필요한 연산의 수

        int i = N-1;  // n-1부터 시작,

        while(i > 0) {
            if(3*i <= N) {  // N 범위 안이면
                dp[i] = Math.min(dp[2*i], Math.min(dp[3*i], dp[i+1])) + 1;  // 3 혹은 2로 나누어 떨어지는 값 중 최소
            } else if(2*i <= N) {
                dp[i] = Math.min(dp[2*i], dp[i+1]) + 1;  // 3 혹은 2로 나누어 떨어지는 값 중 최소
            } else {
                // 1 빼는 경우
                dp[i] = dp[i+1] + 1;
            }
            i--;
        }

        System.out.println(dp[1]);
    }
}